programs/perpetuals/tests/anchor/basic.ts

```diff
@@ -408,6 +419,84 @@ describe("perpetuals", () => {
     expect(JSON.stringify(oracle)).to.equal(JSON.stringify(oracleExpected));
   });
 
+  it("setCustomOraclePricePermissionless", async () => {
+    await tc.setCustomOraclePricePermissionless(
+      permissionlessOracleKeypair,
+      500,
+      tc.custodies[0]
+    );
+
+    let oracle = await tc.program.account.customOracle.fetch(
+      tc.custodies[0].oracleAccount
+    );
+    let oracleExpected = {
+      price: new BN(500000),
+      expo: -3,
+      conf: new BN(0),
+      ema: new BN(500000),
+      publishTime: oracle.publishTime,
+    };
+    expect(JSON.stringify(oracle)).to.equal(JSON.stringify(oracleExpected));
+
+    // Updating the permissionless price oracle with an older publish time should no-op.
+    await tc.setCustomOraclePricePermissionless(
+      permissionlessOracleKeypair,
+      400,
+      tc.custodies[0],
+      tc.getTime() - 20
+    );
+    oracle = await tc.program.account.customOracle.fetch(
+      tc.custodies[0].oracleAccount
+    );
+    // Oracle's value is still 500 instead of the attempted 400.
+    expect(JSON.stringify(oracle)).to.equal(JSON.stringify(oracleExpected));
+
+    // after test, set price back to the expected for other test cases.
+    await tc.setCustomOraclePricePermissionless(
+      permissionlessOracleKeypair,
+      123,
+      tc.custodies[0],
+      tc.getTime() + 1
+    );
+  });
+
+  it("setCustomOraclePricePermissionless Errors", async () => {
+    // Attempting to update with a payload signed by an unknown key should fail.
+    let unknownOraclePrivateKey = Keypair.generate();
+    await tc.ensureFails(
+      tc.setCustomOraclePricePermissionless(
+        unknownOraclePrivateKey,
+        100,
+        tc.custodies[1]
+      )
+    );
+
+    // Sending the permissionless update without signature verification should fail.
+    await tc.ensureFails(
+      tc.setCustomOraclePricePermissionless(
+        permissionlessOracleKeypair,
+        100,
+        tc.custodies[1],
+        null,
+        true
+      )
+    );
+
+    // Sending the permissionless update without malformed message should fail.
```

Comment:
could you also add a test that sends a bogus signature?

Reply:
done! added above and renamed to `bogusKeypair`